First draft of core version policy

jtcohen6 opened the pull request:
resolves #1055

Super rough first draft! Made especially rough by the fact that I can't build the docs site locally (need to un-brick my whole `npm` situation)

I'd like feedback on:
- Where should this page go?
- Whether the semantic versioning intro is needed — I see it in other OSS companies' pages, and it's pretty critical to understanding what we mean in the following sections
- Nomenclature: "Active Support"? "End of Support" vs. "Security Support Until"?
- How to best represent/visualize the timeline at the end (knowing that a lot of it is indicative only)

## Pre-release docs
Is this change related to an unreleased version of dbt?
- [ ] Yes: please
    - update the base branch to `next`
    - add Changelog components: `<Changelog>[New/Changed] in v0.x.0</Changelog>`
    - add links to the "New and changed documentation" section of the latest [Migration Guide](../website/docs/docs/guides/migration-guide)
- [x] No: please ensure the base branch is `current`
- [ ] Unsure: we'll let you know!

## Checklist
If you added new pages (delete if not applicable):
- [x] The page has been added to `website/sidebars.js`
- [x] The new page has a unique filename

jtcohen6 committed:
First draft of core version policy

runleonarun commented on website/docs/docs/core-versions.md:
```diff
@@ -0,0 +1,52 @@
+See also:
+- [Installing dbt Core](dbt-cli/install/overview)
+- [require-dbt-version](require-dbt-version) + [dbt_version](dbt_version)
+- In dbt Cloud: [choosing](cloud-choosing-a-dbt-version) and [upgrading](cloud-upgrading-dbt-versions) dbt Core versions
```
```suggestion
- To learn how you can use dbt Core versions in dbt Cloud, see "[Choosing a dbt Core version](cloud-choosing-a-dbt-version)" and "[Upgrading dbt Core versions in Cloud](cloud-upgrading-dbt-versions)."
```

runleonarun commented on website/docs/docs/core-versions.md:
```diff
@@ -0,0 +1,52 @@
+---
+title: "Understanding dbt Core Versions"
```
```suggestion
title: "Understanding dbt Core versions"
```

runleonarun commented on website/docs/docs/core-versions.md:
```diff
@@ -0,0 +1,52 @@
+---
+title: "Understanding dbt Core Versions"
+id: "core-versions"
+---
```
Open to changes here. Just a 1st shot.

```suggestion
description: "Learn about the semantic versioning dbt Core uses and when we end support for versions."
---
```

runleonarun commented on website/docs/docs/core-versions.md:
```diff
@@ -0,0 +1,52 @@
+
+See also:
+- [Installing dbt Core](dbt-cli/install/overview)
+- [require-dbt-version](require-dbt-version) + [dbt_version](dbt_version)
```
```suggestion
- To restrict your project to only work with a range of dbt Core versions or find the currently running dbt Core version, see "[require-dbt-version](require-dbt-version)" and "[dbt_version](dbt_version)."
```

runleonarun commented on website/docs/docs/core-versions.md:
```diff
@@ -0,0 +1,52 @@
+---
+
+See also:
+- [Installing dbt Core](dbt-cli/install/overview)
```
```suggestion
- To learn about the different ways you can install dbt Core, see "[How to install dbt Core](dbt-cli/install/overview)."
```

runleonarun commented on website/docs/docs/core-versions.md:
```diff
@@ -0,0 +1,52 @@
+### Semantic versioning
+
+dbt Core is released in versions, according to [semantic versioning](https://semver.org/). There are three types of version releases:
+
```
```suggestion

- **Major versions** To date, dbt Core has had one major version release (v1.0.0). When released, v2.0.0 will introduce new features and break backwards compatibility for functionality that has been deprecated.
```

runleonarun commented on website/docs/docs/core-versions.md:
```diff
@@ -0,0 +1,52 @@
+- **Minor versions**, also called "feature releases," include a mix of net-new features and changes to existing capabilities that are backwards-compatible with previous minor versions. They will not break code in your project that relies on documented functionality.
+- **Patch versions**, also called "maintenance releases," include fixes _only_. These fixes could be improvements to new features, needed to restore previous behavior (from unintended regressions), or critical for security/installation. We are judicious about which fixes are included in patch releases, to minimize the surface area of changes.
+
+To date, dbt Core has had one **major version** release: v1.0.0. In the future, when v2.0.0 is released, it will introduce new features and break backwards compatibility for functionality that has been deprecated.
```
Per my comment above, can we move this as the 1st bullet point in the list?
```suggestion
```

runleonarun commented on website/docs/docs/core-versions.md:
```diff
@@ -0,0 +1,52 @@
+
+### Semantic versioning
+
+dbt Core is released in versions, according to [semantic versioning](https://semver.org/). There are three types of version releases:
```
```suggestion
dbt Core is released in versions, according to [semantic versioning](https://semver.org/), which means three types of version releases:
```

runleonarun commented on website/docs/docs/core-versions.md:
```diff
@@ -0,0 +1,52 @@
+- [require-dbt-version](require-dbt-version) + [dbt_version](dbt_version)
+- In dbt Cloud: [choosing](cloud-choosing-a-dbt-version) and [upgrading](cloud-upgrading-dbt-versions) dbt Core versions
+
+### Semantic versioning
```
Let's move these up to H2 and if we think H2 is too big, maybe I should open an issue to deal with it in the CSS.
```suggestion
## Semantic versioning
```

runleonarun commented on website/docs/docs/core-versions.md:
```diff
@@ -0,0 +1,52 @@
+
+To date, dbt Core has had one **major version** release: v1.0.0. In the future, when v2.0.0 is released, it will introduce new features and break backwards compatibility for functionality that has been deprecated.
+
+### Expectations
```
```suggestion
## Expectations
```

runleonarun commented on website/docs/docs/core-versions.md:
```diff
@@ -0,0 +1,52 @@
+
+### Expectations
+
+**Upgrading to patches:** We expect users to upgrade to patches as soon as they're available. When we refer to a "minor version" of dbt Core, such as v1.0 or v1.0.x, we are always referring to _the latest patch release of that version_. We highly encourage you to structure your development and production environments such that you can always install the latest patch.
```
I think this section might be more glanceable if we use H3's. 
```suggestion
### Upgrading to patches 
We expect users to upgrade to patches as soon as they're available. When we refer to a "minor version" of dbt Core, such as v1.0 or v1.0.x, we are always referring to _the latest patch release of that version_. We highly encourage you to structure your development and production environments such that you can always install the latest patch.
```

runleonarun commented on website/docs/docs/core-versions.md:
```diff
@@ -0,0 +1,52 @@
+
+**Upgrading to patches:** We expect users to upgrade to patches as soon as they're available. When we refer to a "minor version" of dbt Core, such as v1.0 or v1.0.x, we are always referring to _the latest patch release of that version_. We highly encourage you to structure your development and production environments such that you can always install the latest patch.
+
+**Upgrading minor versions:** While we do not expect users to immediately upgrade to minor versions as soon as they're available, there will always be features and some fixes that are only available for users of the latest minor version.
```
```suggestion
### Upgrading minor versions 
While we do not expect users to immediately upgrade to minor versions as soon as they're available, there will always be features and some fixes that are only available for users of the latest minor version.
```

runleonarun commented on website/docs/docs/core-versions.md:
```diff
@@ -0,0 +1,52 @@
+
+Note that, for **adapter versions**, dbt-core and adapter plugins coordinate minor releases. Patch version numbers may differ, and will be released independently. As such, you may find you're using `dbt-core==1.0.1` with `dbt-snowflake==1.0.0` — that's ok! You should always install and use the latest available patch for all `dbt-*` software.
+
+**Trying prereleases:** All dbt Core versions are available as _prereleases_ before the final release. "Release candidates" are available for testing, in production-like environments, two weeks before the final release. For minor versions, we also aim to release one or more "betas," which include new features and invite community feedback, 4+ weeks before the final release. It is in your interest ...
```
```suggestion
## Trying prereleases 
All dbt Core versions are available as _prereleases_ before the final release. "Release candidates" are available for testing, in production-like environments, two weeks before the final release. For minor versions, we also aim to release one or more "betas," which include new features and invite community feedback, 4+ weeks before the final release. It is in your interest to help us test prereleases—we need your help!
```

runleonarun replied:
```suggestion
### Trying prereleases 
All dbt Core versions are available as _prereleases_ before the final release. "Release candidates" are available for testing, in production-like environments, two weeks before the final release. For minor versions, we also aim to release one or more "betas," which include new features and invite community feedback, 4+ weeks before the final release. It is in your interest to help us test prereleases—we need your help!
```

runleonarun commented on website/docs/docs/core-versions.md:
```diff
@@ -0,0 +1,52 @@
+
+**Trying prereleases:** All dbt Core versions are available as _prereleases_ before the final release. "Release candidates" are available for testing, in production-like environments, two weeks before the final release. For minor versions, we also aim to release one or more "betas," which include new features and invite community feedback, 4+ weeks before the final release. It is in your interest ...
+
+### Timeline + support
```
```suggestion
## Timeline and support
```

runleonarun commented on website/docs/docs/core-versions.md:
```diff
@@ -0,0 +1,52 @@
+
+### Timeline + support
+
+_Prior to v1.0_
```
```suggestion
### Versioning prior to v1.0
```

runleonarun commented on website/docs/docs/core-versions.md:
```diff
@@ -0,0 +1,52 @@
+- We are no longer releasing new patches for minor versions prior to v1.0
+- As of June 30, 2022, dbt Cloud will remove support for dbt Core versions older than v1.0
+
+_Starting with v1.0_
```
```suggestion
### Versioning starting with v1.0
```

runleonarun replied:
```suggestion
### Versioning starting with v1.0
```

runleonarun commented on website/docs/docs/core-versions.md:
```diff
@@ -0,0 +1,52 @@
+
+_Starting with v1.0_
+
+**Past versions:** Minor versions will be supported for one year (12 months) from the date of their initial release. _This is a definite commitment._ Ongoing support includes:
```
```suggestion
#### Past versions
Minor versions will be supported for one year (12 months) from the date of their initial release. _This is a definite commitment._ Ongoing support includes:
```

runleonarun commented on website/docs/docs/core-versions.md:
```diff
@@ -0,0 +1,52 @@
+- **Ongoing patches:** We will continue to release new patch versions that include fixes. When a newer minor version is available, we will end Active Support; subsequent patches will be limited to critical fixes related to security and installation. After End of Support, no new patches will be released.
+- **Availability in dbt Cloud:** While a minor version is officially supported, you can continue to use it in dbt Cloud. We will notify you as its End of Support is approaching. On the day it reaches End of Support, we will upgrade you to the subsequent minor version.
+
+**Future versions:** We aim to release a new minor "feature" every 3 months. _This is an indicative timeline ONLY._ For the latest information about upcoming releases, including their planned release dates and which features and fixes might be included in each, see [the GitHub milestones in the dbt-core repository](https://github.com/dbt-labs/dbt-core/milestones).
```
```suggestion
#### Future versions

We aim to release a new minor "feature" every 3 months. _This is an indicative timeline ONLY._ For the latest information about upcoming releases, including their planned release dates and which features and fixes might be included in each, see [the GitHub milestones in the dbt-core repository](https://github.com/dbt-labs/dbt-core/milestones).
```

runleonarun reviewed (commented):
@jtcohen6 I made some structural suggestions and a few wording ones. Feel free to let me know if there's a better way to get my ideas across!

runleonarun commented:
@jtcohen6 I neglected to address the feedback you asked for! Here are my thoughts:
> Where should this page go?
 
I think starting with the current placement makes sense. We will need to add more to the Intro to dbt about how Core and Cloud are related, but we need to do that anyway.
 
> Whether the semantic versioning intro is needed — I see it in other OSS companies' pages, and it's pretty critical to understanding what we mean in the following sections
  
I like the semantic intro and that you show how it looks for us
> Nomenclature: "Active Support"? "End of Support" vs. "Security Support Until"?

I was leaning towards End of Support, but I think "Active Support" might actually be clearer and then we don't have to be married to a date. You have both in the table. Were you considering removing one or showing both? We could add the date to the "Active support until" column when it's solidified?

> How to best represent/visualize the timeline at the end (knowing that a lot of it is indicative only)

I really like how PHP does it, especially the calendar: https://www.php.net/supported-versions.php. I like that they link to the release branches too. It would be cool if we could insert the GitHub milestones into the docs, then we don't have to update both places.

jtcohen6 committed:
PR feedback

jtcohen6 commented:
Thank you for the thorough review @runleonarun!! I've addressed your comments (all spot on), and took a second pass while keeping the same overall structure.

One thing I have eyes on: how to coordinate information (and avoid duplicating too much) between this page and the "Choosing a dbt version" for dbt Cloud.

>> Nomenclature: "Active Support"? "End of Support" vs. "Security Support Until"?

> I was leaning towards End of Support, but I think "Active Support" might actually be clearer and then we don't have to be married to a date. You have both in the table. Were you considering removing one or showing both? We could add the date to the "Active support until" column when it's solidified?

Ah, I intended different things here! Added some more text that hopefully clarifies the distinction between "Active" and "Security" support periods.

> It would be cool if we could insert the GitHub milestones into the docs, then we don't have to update both places.

Yes... let's think more on this.

jtcohen6 committed:
Merge branch 'current' into feature/core-version-policy

drewbanin commented:
I just removed myself from review! @jtcohen6, let me know if there's anything in particular in here that you'd like me to dig into / comment on :)

runleonarun commented:
Hey @jtcohen6 this looks good! I will go through and see if we can cut anything to reduce its length. I also have some ideas for moving sections around to bring the most important stuff up higher, and more context down lower. I will do this as soon as I can!

> One thing I have eyes on: how to coordinate information (and avoid duplicating too much) between this page and the "Choosing a dbt version" for dbt Cloud.

I started capturing places with duplicate content in [this issue](https://github.com/dbt-labs/docs.getdbt.com/issues/1126) so when we release the dbt Core versioning, we will also get the reusable code snippets so we can more easily write content once and share it in multiple places. 

I wonder if the "Availability in dbt Cloud" should live in the Cloud docs (maybe with a different title)? We can link to it from here? The version policy page is pretty long and little things like moving this content this could help people consume the information. This is one of the things I will move around and you can yay or nay it. 

I noticed the code comments show up on staging. I wonder if we need to use three dashes `<!---` instead? I can play with this but haven't had time to just yet. 

>  <!-- TODO: Is this true? Do we want to say it here? No breaking changes to project code after v1.0, so this feels like a reasonable approach for us to take. --->

jtcohen6 committed:
Verbal feedback

jtcohen6 commented:
@runleonarun I pushed the structural changes we discussed live yesterday. I think it's ready for one last read-through and language check

jtcohen6 commented on website/docs/docs/core-versions.md:
```diff
@@ -0,0 +1,96 @@
+---
+title: "Understanding dbt Core versions"
```
Should this just be `Understanding dbt versions`, or `dbt Core versions`? I see it's the widest thing on the default sidebar

runleonarun replied:
@jtcohen6 This is a great point. Maybe "About dbt Core versions?" I will be making those other titles more screen-reader friendly so this won't be the only longish one.

Commit:
Merge branch 'current' into feature/core-version-policy

Commit:
Merge branch 'current' into feature/core-version-policy

Commit:
condensing the ideas a teeny bit

runleonarun commented:
OK @jtcohen6 I think we can publish if you are OK with my changes!

runleonarun reviewed (approved):
:shipit:

jtcohen6 committed:
Merge branch 'current' into feature/core-version-policy
